torch/csrc/jit/source_range_serialization.cpp

```diff
@@ -0,0 +1,135 @@
+#include <torch/csrc/jit/source_range_serialization.h>
+
+#include <ATen/core/ivalue.h>
+#include <torch/csrc/jit/pickler.h>
+
+namespace torch {
+namespace jit {
+
+class SourceRangeSerializer {
+ public:
+  // Serialize SourceRange as Tuple[SourceType, int, int]
+  // where SourceType = Tuple[str, Optional[str], int, List[int]],
+  // the serialized form of Source
+  c10::IValue serialize(const SourceRange& sr);
+
+ private:
+  // Serialize Source as Tuple[str, Optional[str], int, List[int]]
+  // This caches serialized sources, since many SourceRanges can
+  // refer to the same one.
+  c10::IValue serialize_source(const std::shared_ptr<Source>& s);
+
+  std::unordered_map<std::shared_ptr<Source>, c10::IValue> serialized_sources;
+};
+
+class SourceRangeDeserializer {
+ public:
+  SourceRange deserialize(const c10::IValue& iv) {
+    auto tup_elems = iv.toTuple()->elements();
+    TORCH_INTERNAL_ASSERT(tup_elems.size() == 3);
+    std::shared_ptr<Source> source_ = deserialize_source(tup_elems[0]);
+    int64_t start_ = tup_elems[1].toInt();
+    int64_t end_ = tup_elems[2].toInt();
+    return SourceRange(source_, start_, end_);
+  }
+
+ private:
+  std::shared_ptr<Source> deserialize_source(const c10::IValue& iv) {
+    // TODO: cache these?
```

Comment:
Yeah, we need to do this one because otherwise it is copy the entire source string over and over again.

Reply:
The key is just the pointer that is being stored inside in the IValue. I think we have a method that accesses it (see the pickler, which does something similar).